### PR TITLE
[READY] Add field to customize Jedi settings

### PR DESCRIPTION
**Pull request by micbou**
This PR allows customization of [the global Jedi settings](https://jedi.readthedocs.io/en/latest/docs/settings.html) by specifying the optional `settings` field in a request. This field is available for all endpoints except the `healthy` and `ready` ones where it is not needed.

When the `settings` field is given, the global Jedi settings are only modified during the request. At the end of the request, the settings are reverted to their default values. This makes it possible to use JediHTTP with different sets of settings; one for each Python project for instance.

I added a test that is a practical example of how this new feature can be useful: by setting the `additional_dynamic_modules` option, we can find references of a symbol that are not in the current module or the imported modules. This is the option that we will use to improve the `GoToReferences` command and to implement [the `RefactorRename` command in ycmd](https://github.com/Valloric/ycmd/pull/425).

---
This change is [Reviewable](https://reviewable.io/reviews/vheon/jedihttp/25)

### CODE REVIEW

**Comment by lithammer** [comment]
Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

_[jedihttp/compatibility.py, line 68 at r1](https://reviewable.io:443/reviews/vheon/jedihttp/25#-KWsAYrUYMhGHTcx8xbu:-KWsAYrUYMhGHTcx8xbv:b-vnb7ik) ([raw file](https://github.com/vheon/jedihttp/blob/0b2c889e6d728bd7ba4e03afe7625a5780bd59a6/jedihttp/compatibility.py#L68)):_

> ``` Python
>   # Python 3
>   def iteritems( dictionary ):
>     return iter( dictionary.items() )
> ```

`items()` should already be an iterator here, no need for `iter()`.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/vheon/jedihttp/25)_

**Comment by micbou** [comment]
Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_[jedihttp/compatibility.py, line 68 at r1](https://reviewable.io:443/reviews/vheon/jedihttp/25#-KWsAYrUYMhGHTcx8xbu:-KWsG4C2M_gdutmM18Um:b-jfjsqs) ([raw file](https://github.com/vheon/jedihttp/blob/0b2c889e6d728bd7ba4e03afe7625a5780bd59a6/jedihttp/compatibility.py#L68)):_
<details><summary>

<i>Previously, renstrom (Peter Renström) wrote…</i></summary>



> `items()` should already be an iterator here, no need for `iter()`.
> </details>
> I took this code from http://legacy.python.org/dev/peps/pep-0469/. If we remove `iter()`, we don't return the same object as the one returned on Python 2. That means we can't use `iteritems` in the same situations on both Python.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/vheon/jedihttp/25)_

**Comment by vheon** [comment]
I just gave this a quick look and it looks ok; the only thing that came to mind is: does it make sense to make a handle or a command line flags to start a JediHTTP with some settings? I understand that, at least for our use in ycmd, make sense to specify some settings only for a request, like we want to make jedi look for other modules other than the one imported for `GoToReferences` but not for `GoTo` to be more efficient, but does it make sense to specify some settings once and for all without changing settings at every request to change them back at the end of it? I'm going to read the jedi docs more carefully but if you already know the answer would be awesome :stuck_out_tongue_closed_eyes: 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/vheon/jedihttp/25#-:-KWt6x_BGXYHtxByqGsf:bn57yeu)_

**Comment by micbou** [comment]
I can't think of a good reason to do that:
- performance is not a concern: setting at most 36 variables (there are 18 settings) for each request is nothing;
- less convenient and less efficient when dealing with multiple sets of settings because an additional request must be sent to change the settings before the actual request;
- not really simpler for clients to implement: sending once a request to change settings versus adding the `settings` field at each request.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/vheon/jedihttp/25#-:-KWtDbxvlaGcN_H4blN-:b-8rv0xa)_

**Comment by vheon** [comment]
> less convenient and less efficient when dealing with multiple sets of settings because an additional request must be sent to change the settings before the actual request;

I think I express myself poorly, let me try again: Is it a plausible scenario for a client to specify the same settings for all request? If so, wouldn't be convenient to have an API that lets you change the "default" settings for a JediHTTP instance?

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/vheon/jedihttp/25#-:-KWtbyu1qpY6CbBKIOVm:bjoqkfo)_

**Comment by micbou** [comment]
> Is it a plausible scenario for a client to specify the same settings for all request?

Yes.

> If so, wouldn't be convenient to have an API that lets you change the "default" settings for a JediHTTP instance?

Think about how you would implement each approach in the Python completer. In one case, you would add the settings in [`_TranslateRequestForJediHTTP`](https://github.com/Valloric/ycmd/blob/master/ycmd/completers/python/jedi_completer.py#L240-L245). In the other case, you would pass the settings in `_StartServer`. I don't see one being more convenient than the other.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/vheon/jedihttp/25#-:-KWtggNwIEpBhVjkIaRR:btpvap2)_

**Comment by vheon** [comment]
I see that for our use in ycmd is not needed, so if I choose to add another API for this I'm going to do it myself later :+1: 

I left a couple of comments :)

---

Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

_[jedihttp/handlers.py, line 117 at r1](https://reviewable.io:443/reviews/vheon/jedihttp/25#-KWu0rPi067Az67g7tr4:-KWu0rPi067Az67g7tr5:b-gnmfs) ([raw file](https://github.com/vheon/jedihttp/blob/0b2c889e6d728bd7ba4e03afe7625a5780bd59a6/jedihttp/handlers.py#L117)):_

> ``` Python
>     request_json = request.json
>     with _CustomSettings( request_json ):
>       script = _GetJediScript( request.json )
> ```

I guess we can use `request_json` here too.

---

_[jedihttp/utils.py, line 18 at r1](https://reviewable.io:443/reviews/vheon/jedihttp/25#-KWu1RC1VQk7wiXL8vby:-KWu1RC1VQk7wiXL8vbz:b-qrdq8a) ([raw file](https://github.com/vheon/jedihttp/blob/0b2c889e6d728bd7ba4e03afe7625a5780bd59a6/jedihttp/utils.py#L18)):_

> ``` Python
> 
> 
> try:
> ```

This looks like a duplicate of the one in `compatibility.py`

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/vheon/jedihttp/25#-:-KWtztRgd-Ko9UyN9GHH:bwpwx5y)_

**Comment by micbou** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_[jedihttp/handlers.py, line 117 at r1](https://reviewable.io:443/reviews/vheon/jedihttp/25#-KWu0rPi067Az67g7tr4:-KWuF53vYaMd0-VgHWTg:b-73m6vj) ([raw file](https://github.com/vheon/jedihttp/blob/0b2c889e6d728bd7ba4e03afe7625a5780bd59a6/jedihttp/handlers.py#L117)):_
<details><summary>

<i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>



> I guess we can use `request_json` here too.
> </details>
> Nice catch.

---

_[jedihttp/utils.py, line 18 at r1](https://reviewable.io:443/reviews/vheon/jedihttp/25#-KWu1RC1VQk7wiXL8vby:-KWuEt3aCsxng7pURu37:bss4a2b) ([raw file](https://github.com/vheon/jedihttp/blob/0b2c889e6d728bd7ba4e03afe7625a5780bd59a6/jedihttp/utils.py#L18)):_
<details><summary>

<i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>



> This looks like a duplicate of the one in `compatibility.py`
> </details>
> Removed.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/vheon/jedihttp/25)_

**Comment by vheon** [comment]
:lgtm: @homu r+

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/vheon/jedihttp/25#-:-KWwHcdJemUPA4RLzz7Z:b-2cbojr)_
